graph/src/lib/graph.ts

```diff
@@ -2,37 +2,50 @@
  * Starter LangGraph.js Template
  * Make this code your own!
  */
-import { createReactAgent } from '@langchain/langgraph/prebuilt';
+import { ToolNode } from '@langchain/langgraph/prebuilt';
 import { ChatOpenAI } from '@langchain/openai';
 import { tokensSearch, bebopRate, EvmKit } from '@agentic-chat/tools';
-import { SYSTEM_PROMPT } from '@agentic-chat/utils';
-import { MemorySaver } from '@langchain/langgraph/web';
-import { WalletClient } from 'viem';
+import { END, MemorySaver, MessagesAnnotation, START, StateGraph } from '@langchain/langgraph/web';
+import { AIMessage } from "@langchain/core/messages";
 
-// @ts-expect-error TODO: FIXME maybe
-const env = import.meta?.env ? import.meta.env : process.env;
+const env = process.env;
 
+const tools = [tokensSearch, bebopRate, ...new EvmKit().getTools()]
 const model = new ChatOpenAI({
   modelName: 'gpt-4o-mini',
   temperature: 0,
   openAIApiKey: env.VITE_OPENAI_API_KEY,
-});
+}).bindTools(tools);
 
 // Adds persistence
 const checkpointer = new MemorySaver();
 
-// Create and export the agent
-export const graph = createReactAgent({
-  llm: model,
-  tools: [tokensSearch, bebopRate, ...new EvmKit().getTools()],
-  checkpointer,
-  prompt: SYSTEM_PROMPT,
-});
-
-export const makeDynamicGraph = (walletClient: WalletClient | undefined) =>
-  createReactAgent({
-    llm: model,
-    tools: [tokensSearch, bebopRate, ...new EvmKit(walletClient).getTools()],
-    checkpointer,
-    prompt: SYSTEM_PROMPT,
-  });
+function shouldContinue(state: typeof MessagesAnnotation.State): "action" | typeof END {
+  const lastMessage = state.messages[state.messages.length - 1];
+  // If there is no function call, then we finish
+  if (lastMessage && !(lastMessage as AIMessage).tool_calls?.length) {
+      return END;
```

Comment:
Note we return END in `shouldContinue` in this conditional edge so it's not in the graph declarations as an explicit END edge